
## [AnnotSV]()

#### [AnnotSV FAQ](https://lbgi.fr/AnnotSV/faq)
#### [Annotation and Ranking of Human Structural Variations](https://www.lbgi.fr/AnnotSV/ranking)


### AnnotSV parameters
```
$ AnnotSV
AnnotSV 2.1

Copyright (C) 2017-2019 GEOFFROY Veronique

Please feel free to contact me for any suggestions or bug reports
email: veronique.geoffroy@inserm.fr

Tcl/Tk version: 8.6

Application name used (defined with the "ANNOTSV" environment variable):
/home/wuzhikun/anaconda3/envs/NanoSV/bin/AnnotSV_2.1


Arguments are missing see help below


COMMAND LINE USAGE

       $ANNOTSV/bin/AnnotSV -SVinputFile 'Path of your VCF or BED input file with SV coordinates' >& AnnotSV.log &


OPTIONS

-bedtools:                      Path of the bedtools local installation

-candidateGenesFile:            Path of a file containing the candidate genes of the user (gene names can be space-separated, tabulation-separated, or line-break-separated) (optional)

-filteredVCFfiles:              Path of the filtered VCF input file(s) with SNV/indel coordinates for compound heterozygotes report (optional)
                                Gzipped VCF files are supported as well as regular expression

-filteredVCFsamples:            To specifiy the sample names from the VCF files defined from the -filtereVCFfiles option 
                                Default: use all samples from the filtered VCF files

-genomeBuild:                   Genome build used
                                Values: GRCh37 (default) or GRCh38

-help:                          More information on the arguments

-metrics:                       Changing numerical values from frequencies to us or fr metrics (e.g. 0.2 or 0,2)
                                Range values: us (default) or fr

-minTotalNumber:                Minimum number of individuals tested to consider a benign SV for the ranking
                                Range values: [100-1000], default = 500

-outputDir:                     Output path name

-outputFile:                    Output path and file name 

-overlap:                       Minimum overlap (%) between the features (DGV, DDD, promoter, TAD…) and the annotated SV to be reported
                                Range values: [0-100], default = 70

-overwrite:                     To overwrite existing output results
                                Values: yes (default) or no

-promoterSize:                  Number of bases upstream from the transcription start site
                                Default = 500

-SVinputFile:                   Path of the input file (VCF or BED) with SV coordinates
                                Gzipped VCF file is supported 

-SVinputInfo:                   To extract the additional SV input fields and insert the data in the output file
                                Range values: 1 (default) or 0

-SVminSize:                     SV minimum size (in bp)
                                Default = 50

-reciprocal:                    Use of a reciprocal overlap between SV and features (only for annotations with features overlapping the SV)
                                Values: no (default) or yes

-svtBEDcol:                     Number of the column describing the SV type (DEL, DUP)
                                Range values: [4-[, default = -1 (value not given)

-typeOfAnnotation:              Description of the types of lines produced by AnnotSV
                                Values: both (default), full or split

-vcfFiles:                      Path of the VCF input file(s) with SNV/indel coordinates used for false positive discovery
                                Use counts of the homozygous and heterozygous variants
                                Gzipped VCF files are supported as well as regular expression

-vcfPASS:                       Boolean. To only use variants from VCF input files that passed all filters during the calling (FILTER column value equal to PASS)
                                Range values: 0 (default) or 1

-vcfSamples:                    To specifiy the sample names from the VCF files defined from the -vcfFiles option
                                Default: use all samples from the VCF files


-txFile:                        Path of a file containing a list of preferred genes transcripts to be used in priority during the annotation 
                                (Preferred genes transcripts names should be tab or space separated)
```


### used databases



```
$ tree RefGene
RefGene
├── GRCh37
│   └── refGene.sorted.bed
└── GRCh38
    └── refGene.sorted.bed


$ tree Genes-based
Genes-based
├── ACMG
│   └── 20181211_ACMG.tsv
├── ClinGen
│   └── 20181217_ClinGenAnnotations.tsv.gz
├── DDD
│   ├── 20181210_DDG2P.tsv.gz
│   └── 20181218_HI.tsv.gz
├── ExAC
│   ├── 20181217_GeneIntolerance.pLI-Zscore.annotations.tsv.gz
│   └── 20181218_ExAC.CNV-Zscore.annotations.tsv.gz
└── OMIM
    ├── 20181126_morbidGenesCandidates.tsv.gz
    ├── 20181126_morbidGenes.tsv.gz
    └── 20181210_OMIMannotations.tsv.gz


$ tree FtIncludedInSV
FtIncludedInSV
├── dbVar_pathogenic_NR_SV
│   ├── GRCh37
│   │   ├── 20190322_dbVar_pathogenic_NR_SV.formatted.sorted.bed
│   │   └── 20190322_dbVar_pathogenic_NR_SV.header.tsv
│   └── GRCh38
│       ├── 20190322_dbVar_pathogenic_NR_SV.formatted.sorted.bed
│       └── 20190322_dbVar_pathogenic_NR_SV.header.tsv
├── GeneHancer
│   ├── GRCh37
│   └── GRCh38
├── Promoter
│   ├── GRCh37
│   │   └── promoter.500bp.sorted.bed
│   └── GRCh38
│       └── promoter.500bp.sorted.bed
└── TAD
    ├── GRCh37
    │   └── 20171024_boundariesTAD.sorted.bed
    └── GRCh38
        └── 20171127_boundariesTAD.sorted.bed


$ tree SVincludedInFt
SVincludedInFt
├── 1000g
│   ├── GRCh37
│   │   └── 20190318_1000g_SV.sorted.bed
│   └── GRCh38
│       └── 20190318_1000g_SV.sorted.bed
├── DDD
│   └── GRCh37
│       └── 20190318_DDD_population_cnv.sorted.bed
├── DGV
│   ├── GRCh37
│   │   ├── 20190322_DGV.GS_annotations.sorted.bed
│   │   └── 20190322_DGV_samplesInStudies.tsv
│   └── GRCh38
│       ├── 20190322_DGV.GS_liftoverGRCh38_annotations.sorted.bed
│       └── 20190322_DGV_samplesInStudies.tsv
├── gnomAD
│   ├── GRCh37
│   │   ├── 20190402_gnomAD_DEL_SV.sorted.bed
│   │   ├── 20190402_gnomAD_DUP_SV.sorted.bed
│   │   ├── 20190402_gnomAD_INS_SV.sorted.bed
│   │   └── 20190402_gnomAD_INV_SV.sorted.bed
│   └── GRCh38
└── IMH
    ├── GRCh37
    │   ├── 20190404_IMH_DEL_SV.sorted.bed
    │   ├── 20190404_IMH_DUP_SV.sorted.bed
    │   ├── 20190404_IMH_INS_SV.sorted.bed
    │   └── 20190404_IMH_INV_SV.sorted.bed
    └── GRCh38
        ├── 20190405_IMH_DEL_SV.sorted.bed
        ├── 20190405_IMH_DUP_SV.sorted.bed
        ├── 20190405_IMH_INS_SV.sorted.bed
        └── 20190405_IMH_INV_SV.sorted.bed


$ tree BreakpointsAnnotations
BreakpointsAnnotations
├── GCcontent
│   ├── GRCh37
│   │   ├── GRCh37_chromFa.chromSizes
│   │   ├── GRCh37_chromFa.fasta
│   │   └── GRCh37_chromFa.fasta.fai
│   └── GRCh38
│       ├── GRCh38_chromFa.chromSizes
│       ├── GRCh38_chromFa.fasta
│       └── GRCh38_chromFa.fasta.fai
└── Repeat
    ├── GRCh37
    │   └── 20190322_Repeat.sorted.bed
    └── GRCh38
        └── 20190322_Repeat.sorted.bed
```





### run AnnotSV
```
$ /home/wuzhikun/anaconda3/envs/NanoSV/bin/AnnotSV_2.1/bin/AnnotSV -bedtools /home/wuzhikun/anaconda3/envs/WGS/bin/bedtools   -SVinputFile  SRR62_CNV.vcf -genomeBuild GRCh38 -outputDir /home/wuzhikun/Project/WGSSomatic/CNVnator/Anno -outputFile SRR62  -overlap 50 -promoterSize 500 -reciprocal yes
AnnotSV 2.1

Copyright (C) 2017-2019 GEOFFROY Veronique

Please feel free to contact me for any suggestions or bug reports
email: veronique.geoffroy@inserm.fr

Tcl/Tk version: 8.6

Application name used (defined with the "ANNOTSV" environment variable):
/home/wuzhikun/anaconda3/envs/NanoSV/bin/AnnotSV_2.1


...downloading the configuration data (June 05 2019 - 10:00)
    ...configuration data by default
    ...configuration data from /home/wuzhikun/anaconda3/envs/NanoSV/bin/AnnotSV_2.1/configfile
    ...configuration data given in arguments
    ...checking configuration data and files


WARNING: No GeneHancer annotations available.
(Please, see in the README file how to add these annotations. Users need to contact the GeneCards team.)

    ******************************************
    AnnotSV has been run with these arguments:
    ******************************************
    -SVinputFile SRR62_CNV.vcf
    -SVinputInfo 1
    -SVminSize 50
    -bedtools /home/wuzhikun/anaconda3/envs/WGS/bin/bedtools
    -genomeBuild GRCh38
    -metrics us
    -minTotalNumber 500
    -outputDir /home/wuzhikun/Project/WGSSomatic/CNVnator/Anno
    -outputFile SRR62.tsv
    -overlap 50
    -overwrite yes
    -promoterSize 500
    -reciprocal yes
    -svtBEDcol -1
    -vcfPASS 0
    ******************************************


...listing of the annotations to realized (June 05 2019 - 10:01)
    ...refGene annotation
    (with /home/wuzhikun/anaconda3/envs/NanoSV/bin/AnnotSV_2.1/Annotations/RefGene/GRCh38/refGene.sorted.bed)
    ...Genes-based annotations
        ...20181211_ACMG.tsv
        (59 gene identifiers and 1 annotations columns: ACMG)
        ...20181217_ClinGenAnnotations.tsv.gz
        (1362 gene identifiers and 2 annotations columns: HI_CGscore, TriS_CGscore)
        ...20181210_DDG2P.tsv.gz
        (1788 gene identifiers and 5 annotations columns: DDD_status, DDD_mode, DDD_consequence, DDD_disease, DDD_pmids)
        ...20181218_HI.tsv.gz
        (19124 gene identifiers and 1 annotations columns: HI_DDDpercent)
        ...20181217_GeneIntolerance.pLI-Zscore.annotations.tsv.gz
        (18241 gene identifiers and 3 annotations columns: synZ_ExAC, misZ_ExAC, pLI_ExAC)
        ...20181218_ExAC.CNV-Zscore.annotations.tsv.gz
        (15673 gene identifiers and 3 annotations columns: delZ_ExAC, dupZ_ExAC, cnvZ_ExAC)
        ...20181126_morbidGenes.tsv.gz
        (10586 gene identifiers and 1 annotations columns: morbidGenes)
        ...20181126_morbidGenesCandidates.tsv.gz
        (3040 gene identifiers and 1 annotations columns: morbidGenesCandidates)
        ...20181210_OMIMannotations.tsv.gz
        (13736 gene identifiers and 3 annotations columns: Mim Number, Phenotypes, Inheritance)
    ...Annotations with features overlapping the SV
        ...DGV Gold Standard frequency annotation
        ...1000g frequency annotation
        ...Ira M. Hall's lab frequency annotation
    ...Annotations with features overlapped with the SV
        ...Promoters annotation
        ...dbVar_pathogenic_NR_SV annotation
        ...TAD annotation
    ...Breakpoints annotations
        ...GC content annotation
        ...Repeat annotation


...annotation in progress (June 05 2019 - 10:01)
-- GCcontentAnnotation, nuc --
/home/wuzhikun/anaconda3/envs/WGS/bin/bedtools nuc -fi /home/wuzhikun/anaconda3/envs/NanoSV/bin/AnnotSV_2.1/Annotations/BreakpointsAnnotations/GCcontent/GRCh38/GRCh38_chromFa.fasta -bed /home/wuzhikun/Project/WGSSomatic/CNVnator/Anno/20190605_AnnotSV_inputSVfile.formatted.sorted.breakpoints.bed > /home/wuzhikun/Project/WGSSomatic/CNVnator/Anno/20190605_AnnotSV_inputSVfile.formatted.sorted.GCcontent.txt
Feature (11:135086900-135087100) beyond the length of 11 size (135086622 bp).  Skipping.
Feature (12:133275400-133275600) beyond the length of 12 size (133275309 bp).  Skipping.
Feature (13:114364400-114364600) beyond the length of 13 size (114364328 bp).  Skipping.
Feature (14:107043900-107044100) beyond the length of 14 size (107043718 bp).  Skipping.
Feature (15:101991400-101991600) beyond the length of 15 size (101991189 bp).  Skipping.
Feature (16:90338400-90338600) beyond the length of 16 size (90338345 bp).  Skipping.
Feature (18:80373400-80373600) beyond the length of 18 size (80373285 bp).  Skipping.
Feature (19:58617900-58618100) beyond the length of 19 size (58617616 bp).  Skipping.
Feature (2:242193900-242194100) beyond the length of 2 size (242193529 bp).  Skipping.
Feature (20:64444400-64444600) beyond the length of 20 size (64444167 bp).  Skipping.
Feature (3:198295900-198296100) beyond the length of 3 size (198295559 bp).  Skipping.
Feature (4:190214900-190215100) beyond the length of 4 size (190214555 bp).  Skipping.
Feature (5:181538400-181538600) beyond the length of 5 size (181538259 bp).  Skipping.
Feature (8:145138900-145139100) beyond the length of 8 size (145138636 bp).  Skipping.
Feature (9:138394900-138395100) beyond the length of 9 size (138394717 bp).  Skipping.
Feature (X:156040900-156041100) beyond the length of X size (156040895 bp).  Skipping.
Feature (Y:57227400-57227600) beyond the length of Y size (57227415 bp).  Skipping.


...Output columns annotation:
    AnnotSV ID; SV chrom; SV start; SV end; SV length; SV type; ID; REF; ALT; QUAL; FILTER; INFO; FORMAT; /home/wuzhikun/Project/WGSSomatic/CNVnator/SRR62_CNV_raw; AnnotSV type; Gene name; NM; CDS length; tx length; location; intersectStart; intersectEnd; DGV_GAIN_IDs; DGV_GAIN_n_samples_with_SV; DGV_GAIN_n_samples_tested; DGV_GAIN_Frequency; DGV_LOSS_IDs; DGV_LOSS_n_samples_with_SV; DGV_LOSS_n_samples_tested; DGV_LOSS_Frequency; 1000g_event; 1000g_AF; 1000g_max_AF; IMH_ID; IMH_AF; IMH_ID_others; promoters; dbVar_event; dbVar_variant; dbVar_status; TADcoordinates; ENCODEexperiments; GCcontent_left; GCcontent_right; Repeats_coord_left; Repeats_type_left; Repeats_coord_right; Repeats_type_right; ACMG; HI_CGscore; TriS_CGscore; DDD_status; DDD_mode; DDD_consequence; DDD_disease; DDD_pmids; HI_DDDpercent; synZ_ExAC; misZ_ExAC; pLI_ExAC; delZ_ExAC; dupZ_ExAC; cnvZ_ExAC; morbidGenes; morbidGenesCandidates; Mim Number; Phenotypes; Inheritance; AnnotSV ranking


...AnnotSV is done with the analysis (June 05 2019 - 10:09)


```

output file:
```
$ tree Anno/
Anno/
└── SRR62.tsv
```


defail of output file
```
$ less  SRR62.tsv

AnnotSV ID  SV chrom    SV start    SV end  SV length   SV type ID  REF ALT QUAL    FILTER  INFO    FORMAT  /home/wuzhikun/Project/WGSSomatic/CNVnator/SRR62_CNV_raw    AnnotSV type    Gene name   NM  CDS length  tx length   location    intersectStart  intersectEnd    DGV_GAIN_IDs    DGV_GAIN_n_samples_with_SV  DGV_GAIN_n_samples_tested   DGV_GAIN_Frequency  DGV_LOSS_IDs    DGV_LOSS_n_samples_with_SV  DGV_LOSS_n_samples_tested   DGV_LOSS_Frequency  1000g_event 1000g_AF    1000g_max_AF    IMH_ID  IMH_AF  IMH_ID_others   promoters   dbVar_event dbVar_variant   dbVar_status    TADcoordinates  ENCODEexperiments   GCcontent_left  GCcontent_right Repeats_coord_left  Repeats_type_left   Repeats_coord_right Repeats_type_right  ACMG    HI_CGscore  TriS_CGscore    DDD_status  DDD_mode    DDD_consequence DDD_disease DDD_pmids   HI_DDDpercent   synZ_ExAC   misZ_ExAC   pLI_ExAC    delZ_ExAC   dupZ_ExAC   cnvZ_ExAC   morbidGenes morbidGenesCandidates   Mim Number  Phenotypes  Inheritance AnnotSV ranking
1_1_80786500_DUP    1   1   80786500    80786499    DUP CNVnator_dup_1  N   <DUP>   .   PASS    END=80786500;SVTYPE=DUP;SVLEN=80786500;IMPRECISE;natorRD=3.90037e+13;natorP1=0;natorP2=0;natorP3=0;natorP4=0;natorQ0=0.0887775  GT:CN   ./1:39003700000000  full    A3GALT2/AADACL3/AADACL4/ACADM/ACAP3/ACOT11/ACOT7/ACTG1P20/ACTL8/ACTRT2/ADGRB2/ADGRL4/ADPRHL2/AGBL4/AGBL4-IT1/AGMAT/AGO1/AGO3/AGO4/AGRN/AGTRAP/AHDC1/AJAP1/AK2/AK4/AK5/AKIRIN1/AKR1A1/AKR7A2/AKR7A3/AKR7L/ALDH4A1/ALG6/ALPL/ANGPTL3/ANGPTL7/ANKRD13C/ANKRD65/ARHGEF10L/ARHGEF16/ARHGEF19/ARID1A/ARMH1/ARTN/ASAP3/ASB17/ATAD3A/ATAD3B/ATAD3C/ATG4C/ATP13A2/ATP5IF1/ATP6V0B/ATPAF1/AUNIP/AURKAIP1/AZIN2/B3GALT6/B4GALT2/BEND5/BEST4/BMP8A/BMP8B/BSDC1/BSND/BTBD19/BTF3L4/C1QA/C1QB/C1QC/C1QTNF12/C1orf109/C1orf122/C1orf127/C1orf141/C1orf158/C1orf159/C1orf167/C1orf174/C1orf185/C1orf195/C1orf210/C1orf216/C1orf232/C1orf50/C1orf87/C1orf94/C8A/C8B/CA6/CACHD1/CALML6/CAMK2N1/CAMTA1/CAMTA1-DT/CAP1/CAPZB/CASP9/CASZ1/CATSPER4/CC2D1B/CCDC163/CCDC17/CCDC24/CCDC27/CCDC28B/CCDC30/CCNL2/CD164L2/CD52/CDA/CDC20/CDC42/CDCA8/CDCP2/CDK11A/CDK11B/CDKN2C/CELA2A/CELA2B/CELA3A/CELA3B/CENPS/CENPS-CORT/CEP104/CEP85/CFAP57/CFAP74/CHD5/CITED4/CLCN6/CLCNKA/CLCNKB/CLDN19/CLIC4/CLSPN/CLSTN1/CMPK1/CNKSR1/CNR2/COA7/COL16A1/COL8A2/COL9A2/CORT/CPLANE2/CPT2/CPTP/CROCC/CROCCP2/CROCCP3/CRYBG2/CRYZ/CSF3R/CSMD2/CSMD2-AS1/CTH/CTNNBIP1/CTPS1/CTRC/CYB5RL/CYP2J2/CYP4A11/CYP4A22/CYP4B1/CYP4X1/CYP4Z1/CYP4Z2P/CZIB/DAB1/DAB1-AS1/DCDC2B/DDI2/DDOST/DDX11L1/DEPDC1/DEPDC1-AS1/DFFA/DFFB/DHCR24/DHDDS/DHRS3/DIO1/DIRAS3/DISP3/DLEU2L/DLGAP3/DLSTP1/DMAP1/DMBX1/DMRTA2/DMRTB1/DNAJB4/DNAJC11/DNAJC16/DNAJC6/DNAJC8/DNALI1/DOCK7/DPH2/DRAXIN/DVL1/E2F2/EBNA1BP2/ECE1/ECHDC2/EDN2/EFCAB14/EFCAB14-AS1/EFCAB7/EFHD2/EIF2B3/EIF3I/EIF4G3/ELAVL4/ELOA/ELOA-AS1/ELOVL1/EMC1/EMC1-AS1/ENO1/ENO1-AS1/EPB41/EPHA10/EPHA2/EPHA8/EPHB2/EPS15/ERI3/ERI3-IT1/ERICH3/ERICH3-AS1/ERMAP/ERRFI1/ESPN/ESPNP/EVA1B/EXO5/EXOSC10/EXOSC10-AS1/EXTL1/EYA3/FAAH/FAAHP1/FAAP20/FABP3/FAF1/FAM110D/FAM131C/FAM138A/FAM138C/FAM138D/FAM138F/FAM151A/FAM167B/FAM183A/FAM229A/FAM41C/FAM43B/FAM76A/FAM87B/FBLIM1/FBXO2/FBXO42/FBXO44/FBXO6/FCN3/FGGY/FGR/FHAD1/FHL3/FLJ37453/FNDC10/FNDC5/FOXD2/FOXD2-AS1/FOXD3/FOXD3-AS1/FOXE3/FOXJ3/FOXO6/FPGT/FPGT-TNNI3K/FUBP1/FUCA1/FYB2/GABRD/GADD45A/GALE/GIPC2/GJA4/GJA9/GJA9-MYCBP/GJB3/GJB4/GJB5/GLIS1/GMEB1/GNB1/GNG12/GNG12-AS1/GNL2/GPATCH3/GPBP1L1/GPN2/GPR153/GPR157/GPR3/GPX7/GRHL3/GRIK3/GUCA2A/GUCA2B/H6PD/HCRTR1/HDAC1/HECTD3/HES2/HES3/HES4/HES5/HEYL/HHLA3/HIVEP3/HMGB4/HMGCL/HMGN2/HNRNPCL1/HNRNPCL2/HNRNPCL3/HNRNPCL4/HNRNPR/HOOK1/HP1BP3/HPCA/HPCAL4/HPDL/HSD52/HSPB11/HSPB7/HSPG2/HTR1D/HTR6/HYI/ICMT/ID3/IFFO2/IFI44/IFI44L/IFI6/IFNLR1/IGSF21/IL12RB2/IL22RA1/IL23R/INPP5B/INSL5/INTS11/IPO13/IPP/IQCC/ISG15/ITGB3BP/JAK1/JUN/KANK4/KAZN/KAZN-AS1/KCNAB2/KCNQ4/KDF1/KDM1A/KDM4A/KDM4A-AS1/KHDRBS1/KIAA0319L/KIAA0754/KIAA1522/KIAA2013/KIF17/KIF1B/KIF2C/KLF17/KLF18/KLHDC7A/KLHL17/KLHL21/KNCN/KPNA6/KTI12/L1TD1/LACTBL1/LAPTM5/LCK/LDLRAD1/LDLRAD2/LDLRAP1/LEPR/LEPROT/LEXM/LHX8/LIN28A/LINC00115/LINC00337/LINC00339/LINC00466/LINC00853/LINC01128/LINC01134/LINC01135/LINC01137/LINC01141/LINC01144/LINC01225/LINC01226/LINC01342/LINC01343/LINC01345/LINC01346/LINC01355/LINC01358/LINC01359/LINC01360/LINC01389/LINC01398/LINC01562/LINC01635/LINC01646/LINC01647/LINC01648/LINC01654/LINC01672/LINC01685/LINC01707/LINC01714/LINC01715/LINC01739/LINC01748/LINC01753/LINC01755/LINC01756/LINC01757/LINC01758/LINC01767/LINC01770/LINC01771/LINC01772/LINC01777/LINC01778/LINC01781/LINC01783/LINC01784/LINC01788/LINC02238/LINC02567/LINC02574/LINC02593/LINC02606/LNCTAM34A/LOC100129534/LOC100132062/LOC100132287/LOC100133331/LOC100288069/LOC100288175/LOC100506730/LOC100506801/LOC100506985/LOC100507564/LOC100507634/LOC100996583/LOC101926944/LOC101926964/LOC101927139/LOC101927342/LOC101927417/LOC101928163/LOC101928303/LOC101928324/LOC101928626/LOC101928728/LOC101929464/LOC101929536/LOC101929592/LOC101929626/LOC101929721/LOC101929901/LOC101929935/LOC102724450/LOC102724659/LOC102725121/LOC102725193/LOC105376805/LOC105378586/LOC105378591/LOC105378614/LOC105378663/LOC105378683/LOC110117498-PIK3R3/LOC112267871/LOC284632/LOC339539/LOC400743/LOC646471/LOC653160/LOC729737/LRP8/LRRC38/LRRC40/LRRC41/LRRC42/LRRC47/LRRC53/LRRC7/LRRIQ3/LSM10/LURAP1/LUZP1/LYPLA2/LZIC/MACF1/MACO1/MAD2L2/MAGOH/MAN1C1/MANEAL/MAP3K6/MAP7D1/MARCKSL1/MASP2/MAST2/MATN1/MATN1-AS1/MDS2/MEAF6/MECR/MED18/MED8/MEGF6/MFAP2/MFN2/MFSD2A/MGC27382/MGC34796/MIB2/MICOS10/MICOS10-NBL1/MIER1/MIGA1/MIIP/MIR101-1/MIR1256/MIR1262/MIR1290/MIR1302-10/MIR1302-11/MIR1302-2/MIR1302-9/MIR186/MIR1976/MIR200A/MIR200B/MIR30C1/MIR30E/MIR3115/MIR3116-1/MIR3116-2/MIR3117/MIR34A/MIR34AHG/MIR3605/MIR3659/MIR3671/MIR3675/MIR378F/MIR3917/MIR3972/MIR4251/MIR4252/MIR4253/MIR4254/MIR4255/MIR429/MIR4418/MIR4420/MIR4421/MIR4422/MIR4422HG/MIR4425/MIR4632/MIR4684/MIR4689/MIR4695/MIR4711/MIR4781/MIR4794/MIR551A/MIR552/MIR5581/MIR5584/MIR5585/MIR5697/MIR6068/MIR6079/MIR6084/MIR6127/MIR6500/MIR6726/MIR6727/MIR6728/MIR6729/MIR6730/MIR6731/MIR6732/MIR6733/MIR6734/MIR6735/MIR6808/MIR6859-1/MIR6859-2/MIR6859-3/MIR6859-4/MIR7156/MIR761/MIR7846/MKNK1/MKNK1-AS1/MMACHC/MMEL1/MMP23A/MMP23B/MOB3C/MORN1/MPL/MROH7/MROH7-TTC4/MRPL20/MRPL20-AS1/MRPL37/MRPS15/MRTO4/MSH4/MST1L/MST1P2/MTF1/MTFR1L/MTHFR/MTMR9LP/MTOR/MTOR-AS1/MUL1/MUTYH/MXRA8/MYCBP/MYCL/MYOM3/MYSM1/NADK/NASP/NBL1/NBPF1/NBPF3/NCDN/NCMAP/NDC1/NDUFS5/NECAP2/NEGR1/NEGR1-IT1/NEXN/NEXN-AS1/NFIA/NFIA-AS1/NFIA-AS2/NFYC/NFYC-AS1/NIPAL3/NKAIN1/NMNAT1/NOC2L/NOL9/NPHP4/NPPA/NPPA-AS1/NPPB/NR0B2/NRDC/NSUN4/NT5C1A/NUDC/OMA1/OPRD1/OR4F16/OR4F29/OR4F3/OR4F5/ORC1/OSBPL9/OSCP1/OTUD3/OXCT2/OXCT2P1/P3H1/P3R3URF/PABPC4/PABPC4-AS1/PADI1/PADI2/PADI3/PADI4/PADI6/PAFAH2/PANK4/PAQR7/PARK7/PARS2/PATJ/PAX7/PCSK9/PDE4B/PDIK1L/PDPN/PDZK1IP1/PEF1/PER3/PERM1/PEX10/PEX14/PGD/PGM1/PHACTR4/PHC2/PHF13/PIGK/PIGV/PIK3CD/PIK3CD-AS1/PIK3CD-AS2/PIK3R3/PIN1P1/PINK1/PINK1-AS/PITHD1/PLA2G2A/PLA2G2C/PLA2G2D/PLA2G2E/PLA2G2F/PLA2G5/PLCH2/PLEKHG5/PLEKHM2/PLEKHN1/PLK3/PLOD1/PLPP3/PNRC2/PODN/POMGNT1/POU3F1/PPCS/PPIE/PPIEL/PPIH/PPP1R8/PPT1/PQLC2/PRAMEF1/PRAMEF10/PRAMEF11/PRAMEF12/PRAMEF13/PRAMEF14/PRAMEF15/PRAMEF17/PRAMEF18/PRAMEF19/PRAMEF2/PRAMEF20/PRAMEF22/PRAMEF25/PRAMEF26/PRAMEF27/PRAMEF33/PRAMEF34P/PRAMEF36P/PRAMEF4/PRAMEF5/PRAMEF6/PRAMEF7/PRAMEF8/PRAMEF9/PRDM16/PRDM16-DT/PRDM2/PRDX1/PRKAA2/PRKCZ/PRKCZ-AS1/PRPF38A/PRXL2B/PSMB2/PTAFR/PTCH2/PTGER3/PTGFR/PTP4A2/PTPRF/PTPRU/PUM1/PUSL1/RAB3B/RAB42/RABGGTB/RAD54L/RAP1GAP/RAVER2/RBBP4/RBP7/RCAN3/RCAN3AS/RCC1/RCC2/RER1/RERE/RERE-AS1/RHBDL2/RHCE/RHD/RIMKLA/RIMS3/RLF/RNF11/RNF186/RNF19B/RNF207/RNF220/RNF223/RNU1-1/RNU1-2/RNU1-3/RNU1-4/RNU11/RNU5D-1/RNU5E-1/RNU5F-1/RNU6-1/RNU6-2/RNU6-7/RNU6-8/RNU6-9/RNVU1-18/ROR1/ROR1-AS1/RPA2/RPE65/RPL11/RPL22/RPS14P3/RPS15AP10/RPS6KA1/RPS8/RRAGC/RSC1A1/RSPO1/RSRP1/RUNX3/S100PBP/SAMD11/SCARNA1/SCARNA21B/SCMH1/SCNN1D/SCP2/SDC3/SDF4/SDHB/SELENON/SERBP1/SERINC2/SESN2/SF3A3/SFN/SFPQ/SGIP1/SH2D5/SH3BGRL3/SH3D21/SHISAL2A/SKI/SKINT1L/SLC1A7/SLC25A33/SLC25A34/SLC25A34-AS1/SLC25A3P1/SLC2A1/SLC2A1-AS1/SLC2A5/SLC2A7/SLC30A2/SLC35D1/SLC35E2A/SLC35E2B/SLC44A5/SLC45A1/SLC5A9/SLC6A9/SLC9A1/SLFNL1/SLFNL1-AS1/SMAP2/SMIM1/SMIM12/SMPDL3B/SNHG12/SNHG3/SNIP1/SNORA110/SNORA16A/SNORA44/SNORA55/SNORA59A/SNORA59B/SNORA61/SNORA63C/SNORA73A/SNORA73B/SNORD103A/SNORD103B/SNORD103C/SNORD128/SNORD145/SNORD160/SNORD167/SNORD38A/SNORD38B/SNORD45A/SNORD45B/SNORD45C/SNORD46/SNORD55/SNORD99/SNRNP40/SPATA21/SPATA6/SPEN/SPOCD1/SPSB1/SRARP/SRM/SRRM1/SRSF10/SRSF11/SRSF4/SSBP3/SSBP3-AS1/SSU72/ST3GAL3/ST6GALNAC3/ST6GALNAC5/STIL/STK40/STMN1/STPG1/STX12/SVBP/SYF2/SYNC/SYTL1/SZRD1/SZT2/SZT2-AS1/TACSTD2/TAF12/TAL1/TARDBP/TAS1R1/TAS1R2/TAS1R3/TCEA3/TCEANC2/TCTEX1D1/TCTEX1D4/TEKT2/TENT5B/TESK2/TEX38/TEX46/TFAP2E/THAP3/THEMIS2/THRAP3/TIE1/TINAGL1/TM2D1/TMCO2/TMCO4/TMEM125/TMEM200B/TMEM201/TMEM222/TMEM234/TMEM240/TMEM269/TMEM274P/TMEM275/TMEM35B/TMEM39B/TMEM50A/TMEM51/TMEM51-AS1/TMEM52/TMEM53/TMEM54/TMEM59/TMEM61/TMEM69/TMEM82/TMEM88B/TNFRSF14/TNFRSF14-AS1/TNFRSF18/TNFRSF1B/TNFRSF25/TNFRSF4/TNFRSF8/TNFRSF9/TNNI3K/TOE1/TP73/TP73-AS1/TPRG1L/TRABD2B/TRAPPC3/TRIM62/TRIM63/TRIT1/TRNAU1AP/TRNP1/TSPAN1/TSSK3/TTC22/TTC34/TTC39A/TTC39A-AS1/TTC4/TTLL10/TUT4/TXLNA/TXNDC12/TXNDC12-AS1/TYW3/UBE2J2/UBE2U/UBE4B/UBIAD1/UBR4/UBXN10/UBXN10-AS1/UBXN11/UQCRH/UQCRHL/UROD/USP1/USP24/USP33/USP48/UTP11/UTS2/VAMP3/VPS13D/VWA1/VWA5B1/WASF2/WASH7P/WDR78/WDTC1/WLS/WNT4/WRAP73/XKR8/YARS/YBX1/YIPF1/YRDC/YTHDF2/ZBTB17/ZBTB40/ZBTB48/ZBTB8A/ZBTB8B/ZBTB8OS/ZC3H12A/ZCCHC17/ZDHHC18/ZFP69/ZFP69B/ZFYVE9/ZMPSTE24/ZMYM1/ZMYM4/ZMYM4-AS1/ZMYM6/ZMYND12/ZNF362/ZNF436/ZNF436-AS1/ZNF593/ZNF683/ZNF684/ZNF691/ZRANB2/ZRANB2-AS1/ZRANB2-AS2/ZSCAN20/ZSWIM5/ZYG11A/ZYG11B/ZZZ3                                0   0   -1      0   0   -1      -1  -1      -1  1:7601937-50381771_INV;1:31594522-94732184_INV  A3GALT2/AADACL3/AADACL4/ACADM/ACAP3/ACOT11/ACOT7/ACTG1P20/ACTL8/ACTRT2/ADGRB2/ADGRL4/ADPRHL2/AGBL4/AGBL4-IT1/AGMAT/AGO1/AGO3/AGO4/AGRN/AGTRAP/AHDC1/AJAP1/AK2/AK4/AK5/AKIRIN1/AKR1A1/AKR7A2/AKR7A3/AKR7L/ALDH4A1/ALG6/ALPL/ANGPTL3/ANGPTL7/ANKRD13C/ANKRD65/ARHGEF10L/ARHGEF16/ARHGEF19/ARID1A/ARMH1/ARTN/ASAP3/ASB17/ATAD3A/ATAD3B/ATAD3C/ATG4C/ATP13A2/ATP5IF1/ATP6V0B/ATPAF1/AUNIP/AURKAIP1/AZIN2/B3GALT6/B4GALT2/BEND5/BEST4/BMP8A/BMP8B/BSDC1/BSND/BTBD19/BTF3L4/C1QA/C1QB/C1QC/C1QTNF12/C1orf109/C1orf122/C1orf127/C1orf141/C1orf158/C1orf159/C1orf167/C1orf174/C1orf185/C1orf195/C1orf210/C1orf216/C1orf232/C1orf50/C1orf87/C1orf94/C8A/C8B/CA6/CACHD1/CALML6/CAMK2N1/CAMTA1/CAMTA1-DT/CAP1/CAPZB/CASP9/CASZ1/CATSPER4/CC2D1B/CCDC163/CCDC17/CCDC24/CCDC27/CCDC28B/CCDC30/CCNL2/CD164L2/CD52/CDA/CDC20/CDC42/CDCA8/CDCP2/CDK11A/CDK11B/CDKN2C/CELA2A/CELA2B/CELA3A/CELA3B/CENPS/CENPS-CORT/CEP104/CEP85/CFAP57/CFAP74/CHD5/CITED4/CLCN6/CLCNKA/CLCNKB/CLDN19/CLIC4/CLSPN/CLSTN1/CMPK1/CNKSR1/CNR2/COA7/COL16A1/COL8A2/COL9A2/CORT/CPLANE2/CPT2/CPTP/CROCC/CROCCP2/CROCCP3/CRYBG2/CRYZ/CSF3R/CSMD2/CSMD2-AS1/CTH/CTNNBIP1/CTPS1/CTRC/CYB5RL/CYP2J2/CYP4A11/CYP4A22/CYP4B1/CYP4X1/CYP4Z1/CYP4Z2P/CZIB/DAB1/DAB1-AS1/DCDC2B/DDI2/DDOST/DDX11L1/DEPDC1/DEPDC1-AS1/DFFA/DFFB/DHCR24/DHDDS/DHRS3/DIO1/DIRAS3/DISP3/DLEU2L/DLGAP3/DLSTP1/DMAP1/DMBX1/DMRTA2/DMRTB1/DNAJB4/DNAJC11/DNAJC16/DNAJC6/DNAJC8/DNALI1/DOCK7/DPH2/DRAXIN/DVL1/E2F2/EBNA1BP2/ECE1/ECHDC2/EDN2/EFCAB14/EFCAB14-AS1/EFCAB7/EFHD2/EIF2B3/EIF3I/EIF4G3/ELAVL4/ELOA/ELOA-AS1/ELOVL1/EMC1/EMC1-AS1/ENO1/ENO1-AS1/EPB41/EPHA10/EPHA2/EPHA8/EPHB2/EPS15/ERI3/ERI3-IT1/ERICH3/ERICH3-AS1/ERMAP/ERRFI1/ESPN/ESPNP/EVA1B/EXO5/EXOSC10/EXOSC10-AS1/EXTL1/EYA3/FAAH/FAAHP1/FAAP20/FABP3/FAF1/FAM110D/FAM131C/FAM138A/FAM138C/FAM138D/FAM138F/FAM151A/FAM167B/FAM183A/FAM229A/FAM41C/FAM43B/FAM76A/FAM87B/FBLIM1/FBXO2/FBXO42/FBXO44/FBXO6/FCN3/FGGY/FGR/FHAD1/FHL3/FLJ37453/FNDC10/FNDC5/FOXD2/FOXD2-AS1/FOXD3/FOXD3-AS1/FOXE3/FOXJ3/FOXO6/FPGT/FPGT-TNNI3K/FUBP1/FUCA1/FYB2/GABRD/GADD45A/GALE/GIPC2/GJA4/GJA9/GJA9-MYCBP/GJB3/GJB4/GJB5/GLIS1/GMEB1/GNB1/GNG12/GNG12-AS1/GNL2/GPATCH3/GPBP1L1/GPN2/GPR153/GPR157/GPR3/GPX7/GRHL3/GRIK3/GUCA2A/GUCA2B/H6PD/HCRTR1/HDAC1/HECTD3/HES2/HES3/HES4/HES5/HEYL/HHLA3/HIVEP3/HMGB4/HMGCL/HMGN2/HNRNPCL1/HNRNPCL2/HNRNPCL3/HNRNPCL4/HNRNPR/HOOK1/HP1BP3/HPCA/HPCAL4/HPDL/HSD52/HSPB11/HSPB7/HSPG2/HTR1D/HTR6/HYI/ICMT/ID3/IFFO2/IFI44/IFI44L/IFI6/IFNLR1/IGSF21/IL12RB2/IL22RA1/IL23R/INPP5B/INSL5/INTS11/IPO13/IPP/IQCC/ISG15/ITGB3BP/JAK1/JUN/KANK4/KAZN/KAZN-AS1/KCNAB2/KCNQ4/KDF1/KDM1A/KDM4A/KDM4A-AS1/KHDRBS1/KIAA0319L/KIAA0754/KIAA1522/KIAA2013/KIF17/KIF1B/KIF2C/KLF17/KLF18/KLHDC7A/KLHL17/KLHL21/KNCN/KPNA6/KTI12/L1TD1/LACTBL1/LAPTM5/LCK/LDLRAD1/LDLRAD2/LDLRAP1/LEPR/LEPROT/LEXM/LHX8/LIN28A/LINC00115/LINC00337/LINC00339/LINC00466/LINC00853/LINC01128/LINC01134/LINC01135/LINC01137/LINC01141/LINC01144/LINC01225/LINC01226/LINC01342/LINC01343/LINC01345/LINC01346/LINC01355/LINC01358/LINC01359/LINC01360/LINC01389/LINC01398/LINC01562/LINC01635/LINC01646/LINC01647/LINC01648/LINC01654/LINC01672/LINC01685/LINC01707/LINC01714/LINC01715/LINC01739/LINC01748/LINC01753/LINC01755/LINC01756/LINC01757/LINC01758/LINC01767/LINC01770/LINC01771/LINC01772/LINC01777/LINC01778/LINC01781/LINC01783/LINC01784/LINC01788/LINC02238/LINC02567/LINC02574/LINC02593/LINC02606/LNCTAM34A/LOC100129534/LOC100132062/LOC100132287/LOC100133331/LOC100288069/LOC100288175/LOC100506730/LOC100506801/LOC100506985/LOC100507564/LOC100507634/LOC100996583/LOC101926944/LOC101926964/LOC101927139/LOC101927342/LOC101927417/LOC101928163/LOC101928303/LOC101928324/LOC101928626/LOC101928728/LOC101929464/LOC101929536/LOC101929592/LOC101929626/LOC101929721/LOC101929901/LOC101929935/LOC102724450/LOC102724659/LOC102725121/LOC102725193/LOC105376805/LOC105378586/LOC105378591/LOC105378614/LOC105378663/LOC105378683/LOC110117498-PIK3R3/LOC112267871/LOC284632/LOC339539/LOC400743/LOC646471/LOC653160/LOC729737/LRP8/LRRC38/LRRC40/LRRC41/LRRC42/LRRC47/LRRC53/LRRC7/LRRIQ3/LSM10/LURAP1/LUZP1/LYPLA2/LZIC/MACF1/MACO1/MAD2L2/MAGOH/MAN1C1/MANEAL/MAP3K6/MAP7D1/MARCKSL1/MASP2/MAST2/MATN1/MATN1-AS1/MDS2/MEAF6/MECR/MED18/MED8/MEGF6/MFAP2/MFN2/MFSD2A/MGC27382/MGC34796/MIB2/MICOS10/MICOS10-NBL1/MIER1/MIGA1/MIIP/MIR101-1/MIR1256/MIR1262/MIR1290/MIR1302-10/MIR1302-11/MIR1302-2/MIR1302-9/MIR186/MIR1976/MIR200A/MIR200B/MIR30C1/MIR30E/MIR3115/MIR3116-1/MIR3116-2/MIR3117/MIR34A/MIR34AHG/MIR3605/MIR3659/MIR3671/MIR3675/MIR378F/MIR3917/MIR3972/MIR4251/MIR4252/MIR4253/MIR4254/MIR4255/MIR429/MIR4418/MIR4420/MIR4421/MIR4422/MIR4422HG/MIR4425/MIR4632/MIR4684/MIR4689/MIR4695/MIR4711/MIR4781/MIR4794/MIR551A/MIR552/MIR5581/MIR5584/MIR5585/MIR5697/MIR6068/MIR6079/MIR6084/MIR6127/MIR6500/MIR6726/MIR6727/MIR6728/MIR6729/MIR6730/MIR6731/MIR6732/MIR6733/MIR6734/MIR6735/MIR6808/MIR6859-1/MIR6859-2/MIR6859-3/MIR6859-4/MIR7156/MIR761/MIR7846/MKNK1/MKNK1-AS1/MMACHC/MMEL1/MMP23A/MMP23B/MOB3C/MORN1/MPL/MROH7/MROH7-TTC4/MRPL20/MRPL20-AS1/MRPL37/MRPS15/MRTO4/MSH4/MST1L/MST1P2/MTF1/MTFR1L/MTHFR/MTMR9LP/MTOR/MTOR-AS1/MUL1/MUTYH/MXRA8/MYCBP/MYCL/MYOM3/MYSM1/NADK/NASP/NBL1/NBPF1/NBPF3/NCDN/NCMAP/NDC1/NDUFS5/NECAP2/NEGR1/NEGR1-IT1/NEXN/NEXN-AS1/NFIA/NFIA-AS1/NFIA-AS2/NFYC/NFYC-AS1/NIPAL3/NKAIN1/NMNAT1/NOC2L/NOL9/NPHP4/NPPA/NPPA-AS1/NPPB/NR0B2/NRDC/NSUN4/NT5C1A/NUDC/OMA1/OPRD1/OR4F16/OR4F29/OR4F3/OR4F5/ORC1/OSBPL9/OSCP1/OTUD3/OXCT2/OXCT2P1/P3H1/P3R3URF/PABPC4/PABPC4-AS1/PADI1/PADI2/PADI3/PADI4/PADI6/PAFAH2/PANK4/PAQR7/PARK7/PARS2/PATJ/PAX7/PCSK9/PDE4B/PDIK1L/PDPN/PDZK1IP1/PEF1/PER3/PERM1/PEX10/PEX14/PGD/PGM1/PHACTR4/PHC2/PHF13/PIGK/PIGV/PIK3CD/PIK3CD-AS1/PIK3CD-AS2/PIK3R3/PIN1P1/PINK1/PINK1-AS/PITHD1/PLA2G2A/PLA2G2C/PLA2G2D/PLA2G2E/PLA2G2F/PLA2G5/PLCH2/PLEKHG5/PLEKHM2/PLEKHN1/PLK3/PLOD1/PLPP3/PNRC2/PODN/POMGNT1/POU3F1/PPCS/PPIE/PPIEL/PPIH/PPP1R8/PPT1/PQLC2/PRAMEF1/PRAMEF10/PRAMEF11/PRAMEF12/PRAMEF13/PRAMEF14/PRAMEF15/PRAMEF17/PRAMEF18/PRAMEF19/PRAMEF2/PRAMEF20/PRAMEF22/PRAMEF25/PRAMEF26/PRAMEF27/PRAMEF33/PRAMEF34P/PRAMEF36P/PRAMEF4/PRAMEF5/PRAMEF6/PRAMEF7/PRAMEF8/PRAMEF9/PRDM16/PRDM16-DT/PRDM2/PRDX1/PRKAA2/PRKCZ/PRKCZ-AS1/PRPF38A/PRXL2B/PSMB2/PTAFR/PTCH2/PTGER3/PTGFR/PTP4A2/PTPRF/PTPRU/PUM1/PUSL1/RAB3B/RAB42/RABGGTB/RAD54L/RAP1GAP/RAVER2/RBBP4/RBP7/RCAN3/RCAN3AS/RCC1/RCC2/RER1/RERE/RERE-AS1/RHBDL2/RHCE/RHD/RIMKLA/RIMS3/RLF/RNF11/RNF186/RNF19B/RNF207/RNF220/RNF223/RNU1-1/RNU1-2/RNU1-3/RNU1-4/RNU11/RNU5D-1/RNU5E-1/RNU5F-1/RNU6-1/RNU6-2/RNU6-7/RNU6-8/RNU6-9/RNVU1-18/ROR1/ROR1-AS1/RPA2/RPE65/RPL11/RPL22/RPS14P3/RPS15AP10/RPS6KA1/RPS8/RRAGC/RSC1A1/RSPO1/RSRP1/RUNX3/S100PBP/SAMD11/SCARNA1/SCARNA21B/SCMH1/SCNN1D/SCP2/SDC3/SDF4/SDHB/SELENON/SERBP1/SERINC2/SESN2/SF3A3/SFN/SFPQ/SGIP1/SH2D5/SH3BGRL3/SH3D21/SHISAL2A/SKI/SKINT1L/SLC1A7/SLC25A33/SLC25A34/SLC25A34-AS1/SLC25A3P1/SLC2A1/SLC2A1-AS1/SLC2A5/SLC2A7/SLC30A2/SLC35D1/SLC35E2A/SLC35E2B/SLC44A5/SLC45A1/SLC5A9/SLC6A9/SLC9A1/SLFNL1/SLFNL1-AS1/SMAP2/SMIM1/SMIM12/SMPDL3B/SNHG12/SNHG3/SNIP1/SNORA110/SNORA16A/SNORA44/SNORA55/SNORA59A/SNORA59B/SNORA61/SNORA63C/SNORA73A/SNORA73B/SNORD103A/SNORD103B/SNORD103C/SNORD128/SNORD145/SNORD160/SNORD167/SNORD38A/SNORD38B/SNORD45A/SNORD45B/SNORD45C/SNORD46/SNORD55/SNORD99/SNRNP40/SPATA21/SPATA6/SPEN/SPOCD1/SPSB1/SRARP/SRM/SRRM1/SRSF10/SRSF11/SRSF4/SSBP3/SSBP3-AS1/SSU72/ST3GAL3/ST6GALNAC3/ST6GALNAC5/STIL/STK40/STMN1/STPG1/STX12/SVBP/SYF2/SYNC/SYTL1/SZRD1/SZT2/SZT2-AS1/TACSTD2/TAF12/TAL1/TARDBP/TAS1R1/TAS1R2/TAS1R3/TCEA3/TCEANC2/TCTEX1D1/TCTEX1D4/TEKT2/TENT5B/TESK2/TEX38/TEX46/TFAP2E/THAP3/THEMIS2/THRAP3/TIE1/TINAGL1/TM2D1/TMCO2/TMCO4/TMEM125/TMEM200B/TMEM201/TMEM222/TMEM234/TMEM240/TMEM269/TMEM274P/TMEM275/TMEM35B/TMEM39B/TMEM50A/TMEM51/TMEM51-AS1/TMEM52/TMEM53/TMEM54/TMEM59/TMEM61/TMEM69/TMEM82/TMEM88B/TNFRSF14/TNFRSF14-AS1/TNFRSF18/TNFRSF1B/TNFRSF25/TNFRSF4/TNFRSF8/TNFRSF9/TNNI3K/TOE1/TP73/TP73-AS1/TPRG1L/TRABD2B/TRAPPC3/TRIM62/TRIM63/TRIT1/TRNAU1AP/TRNP1/TSPAN1/TSSK3/TTC22/TTC34/TTC39A/TTC39A-AS1/TTC4/TTLL10/TUT4/TXLNA/TXNDC12/TXNDC12-AS1/TYW3/UBE2J2/UBE2U/UBE4B/UBIAD1/UBR4/UBXN10/UBXN10-AS1/UBXN11/UQCRH/UQCRHL/UROD/USP1/USP24/USP33/USP48/UTP11/UTS2/VAMP3/VPS13D/VWA1/VWA5B1/WASF2/WASH7P/WDR78/WDTC1/WLS/WNT4/WRAP73/XKR8/YARS/YBX1/YIPF1/YRDC/YTHDF2/ZBTB17/ZBTB40/ZBTB48/ZBTB8A/ZBTB8B/ZBTB8OS/ZC3H12A/ZCCHC17/ZDHHC18/ZFP69/ZFP69B/ZFYVE9/ZMPSTE24/ZMYM1/ZMYM4/ZMYM4-AS1/ZMYM6/ZMYND12/ZNF362/ZNF436/ZNF436-AS1/ZNF593/ZNF683/ZNF684/ZNF691/ZRANB2/ZRANB2-AS1/ZRANB2-AS2/ZSCAN20/ZSWIM5/ZYG11A/ZYG11B/ZZZ3    copy_number_gain;copy_number_loss;deletion;duplication  nssv13638713;nssv13638876;nssv13639487;nssv13639504;nssv13639620;nssv13639665;nssv13640035;nssv13640262;nssv13640286;nssv13640535;nssv13640569;nssv13640809;nssv13641247;nssv13641324;nssv13642154;nssv13642778;nssv13643109;nssv13643277;nssv13644070;nssv13646717;nssv13647205;nssv13647265;nssv13647423;nssv13647615;nssv13648455;nssv13648744;nssv13648971;nssv13649175;nssv13649375;nssv13649466;nssv13650373;nssv13650929;nssv13651000;nssv13651574;nssv13651583;nssv13651645;nssv13652240;nssv13652463;nssv13653083;nssv13653198;nssv13653300;nssv13654185;nssv13654496;nssv13654536;nssv13654929;nssv13654953;nssv13655057;nssv13655235;nssv13655807;nssv13656051;nssv13656517;nssv14081997;nssv14082061;nssv14082308;nssv14082318;nssv14082321;nssv14082499;nssv1415016;nssv1415217;nssv1415285;nssv1415327;nssv1415490;nssv1494823;nssv1494829;nssv1494832;nssv1494836;nssv1494842;nssv1495166;nssv1495167;nssv1495168;nssv1495722;nssv1602095;nssv1602221;nssv1602225;nssv1602226;nssv1602237;nssv1602242;nssv1602523;nssv1604112;nssv1604129;nssv1604194;nssv1604674;nssv1608157;nssv1608159;nssv1608217;nssv1608295;nssv1609251;nssv1610127;nssv1610268;nssv1610405;nssv1610445;nssv1610446;nssv1610460;nssv3395311;nssv3395783;nssv3395996;nssv3396407;nssv3396408;nssv3396454;nssv3396489;nssv3396509;nssv3396606;nssv3396607;nssv3396744;nssv3396886;nssv3397008;nssv3397132;nssv3397270;nssv3397381;nssv575320;nssv575321;nssv575658;nssv575868;nssv576063;nssv576287;nssv576295;nssv576311;nssv576338;nssv576346;nssv576394;nssv576618;nssv577125;nssv577127;nssv577128;nssv577129;nssv577130;nssv577131;nssv577133;nssv577134;nssv577135;nssv577136;nssv577138;nssv577139;nssv577140;nssv577141;nssv577142;nssv577143;nssv577144;nssv577150;nssv577152;nssv577154;nssv577155;nssv577156;nssv577157;nssv577158;nssv577160;nssv577161;nssv577162;nssv577164;nssv577165;nssv577166;nssv577167;nssv577168;nssv577171;nssv577174;nssv577176;nssv577179;nssv577180;nssv577183;nssv577184;nssv577185;nssv577186;nssv577187;nssv577188;nssv577189;nssv577190;nssv577191;nssv577194;nssv577196;nssv577197;nssv577198;nssv577199;nssv577200;nssv577201;nssv577202;nssv577204;nssv577205;nssv577206;nssv577207;nssv577208;nssv577209;nssv577210;nssv578501;nssv578502;nssv578504;nssv578506;nssv578507;nssv578508;nssv578509;nssv578512;nssv578515;nssv578517;nssv578518;nssv578519;nssv578520;nssv578523;nssv578524;nssv582122;nssv582220;nssv582298;nssv582348;nssv582353;nssv582441;nssv582753;nssv582807;nssv582846;nssv583014;nssv583149;nssv583260;nssv583303;nssv583874;nssv583880;nssv706308;nssv706592;nssv7487148;nssv7487153;nssv7487214;nssv8639458 Likely pathogenic;Pathogenic    1:10339943-11099943;1:10339943-12099943;1:10379943-11059943;1:10379943-11099943;1:10379943-11939943;1:10379943-12019943;1:10379944-11059943;1:10379944-11099943;1:10419944-11059943;1:10419944-11099943;1:10419944-11939943;1:10419944-12059943;1:10459944-11059943;1:10459944-11099943;1:11019944-11859943;1:11019944-11899943;1:11019944-11939943;1:11019944-12019943;1:11019944-12179943;1:11059944-12019943;1:11059944-12059943 ENCFF039LVJ:ENCFF067UVH:ENCFF139JCA:ENCFF156SZO:ENCFF176FDF:ENCFF183RCR:ENCFF241RXN:ENCFF296RLZ:ENCFF302WDA:ENCFF311BGF:ENCFF444WRK:ENCFF458AHX:ENCFF461GOG:ENCFF572FKA:ENCFF603GAT:ENCFF654ZID:ENCFF716CFF:ENCFF718ETU:ENCFF796ORT:ENCFF822VBC:ENCFF869GJT 0.000000 (101 N)    0.370000            1:80783565-80789663 L1PA7       3                                                       yes yes             5
1_1_80786500_DUP    1   1   80786500    80786499    DUP CNVnator_dup_1  N   <DUP>   .   PASS    END=80786500;SVTYPE=DUP;SVLEN=80786500;IMPRECISE;natorRD=3.90037e+13;natorP1=0;natorP2=0;natorP3=0;natorP4=0;natorQ0=0.0887775  GT:CN   ./1:39003700000000  split   A3GALT2 NM_001080438    1028    14334   txStart-txEnd   33306765    33321098        0   0   -1      0   0   -1      -1  -1      -1                                                                                      61.43   2.27028092535663    1.75809197357458    3.6237402371872e-07 -2.62354415217069   0.668347390481294   -2.47162789907313                       2
1_1_80786500_DUP    1   1   80786500    80786499    DUP CNVnator_dup_1  N   <DUP>   .   PASS    END=80786500;SVTYPE=DUP;SVLEN=80786500;IMPRECISE;natorRD=3.90037e+13;natorP1=0;natorP2=0;natorP3=0;natorP4=0;natorQ0=0.0887775  GT:CN   ./1:39003700000000  split   AADACL3 NM_001103170    1228    12652   txStart-txEnd   12716109    12728760        0   0   -1      0   0   -1      -1  -1      -1                                                                                      92.94   0.0361612141503088  -0.109500896056784  0.000110836921442839    -0.0099947629380129 -2.53125239294469   -2.47162789907313                       2

```


### output file title
```
AnnotSV ID
SV chrom
SV start
SV end
SV length
SV type
ID
REF
ALT
QUAL
FILTER
INFO
FORMAT
/home/wuzhikun/Project/WGSSomatic/CNVnator/SRR62_CNV_raw
AnnotSV type
Gene name
NM
CDS length
tx length
location
intersectStart
intersectEnd
DGV_GAIN_IDs
DGV_GAIN_n_samples_with_SV
DGV_GAIN_n_samples_tested
DGV_GAIN_Frequency
DGV_LOSS_IDs
DGV_LOSS_n_samples_with_SV
DGV_LOSS_n_samples_tested
DGV_LOSS_Frequency
1000g_event
1000g_AF
1000g_max_AF
IMH_ID
IMH_AF
IMH_ID_others
promoters
dbVar_event
dbVar_variant
dbVar_status
TADcoordinates
ENCODEexperiments
GCcontent_left
GCcontent_right
Repeats_coord_left
Repeats_type_left
Repeats_coord_right
Repeats_type_right
ACMG
HI_CGscore
TriS_CGscore
DDD_status
DDD_mode
DDD_consequence
DDD_disease
DDD_pmids
HI_DDDpercent
synZ_ExAC
misZ_ExAC
pLI_ExAC
delZ_ExAC
dupZ_ExAC
cnvZ_ExAC
morbidGenes
morbidGenesCandidates
Mim Number
Phenotypes
Inheritance
AnnotSV ranking
```



### [SV ranking/classification](https://www.lbgi.fr/AnnotSV/ranking)
```
Data:
-   Benign SV from the DGV Gold Standard corresponding to a gain (the ones with DGV_GAIN_Frequency > 1% and with DGV_GAIN_n_samples_tested > 500 (default))
-   Benign SV from the DGV Gold Standard corresponding to a loss (the ones with DGV_LOSS_Frequency > 1% and with DGV_LOSS_n_samples_tested > 500 (default))
-   Pathogenic SV from the dbVar NR-SV dataset
-   pLI scores of each genes from ExAC
-   Haploinsufficiency (HI) and triplosensitivity (TriS) scores from ClinGen
-   Morbid genes from OMIM 
-   Candidate morbid genes from OMIM
-   Candidate genes provided by the user
-   Enhancer and promoter elements from GeneHancer
Criteria:
- Class 1 (benign):
The SV overlaps (>70%) with a benign SV with the same SV type
AND the SV does not overlap with a morbid gene (or its enhancer/promoter)
AND the SV does not overlap with morbid gene candidate (or its enhancer/promoter)
AND the SV does not overlap a candidate gene (or its enhancer/promoter)
- Class 2 (likely benign):
The SV has no overlap OR an overlap ≤70% with a benign SV
AND the SV does not overlap with a morbid gene (or its enhancer/promoter)
AND the SV does not overlap with a morbid gene candidate (or its enhancer/promoter)
AND the SV does not overlap with a candidate gene (or its enhancer/promoter)
- Class 3 (variant of unknown significance):
The SV overlaps a morbid gene candidate (or its enhancer/promoter) (with at least 1bp overlap)
OR the SV overlaps a candidate gene (or its enhancer/promoter) (with at least 1bp overlap)
- Class 4 (likely pathogenic):
The SV overlaps a morbid gene (or its enhancer/promoter) (with at least 1bp)
OR for a loss: the SV overlaps a gene (or its enhancer/promoter) with a pLI_ExAC > 0.9 or with a HI_CGscore value of 3 or 2
OR for a gain: the SV overlaps a gene (or its enhancer/promoter) with a TriS_CGscore value of 3 or 2
- Class 5 (pathogenic):
The SV overlaps a pathogenic SV (with at least 1bp) with the same SV type
```

