

## CNVnator docker

### step 1
```
$ docker run  -v /home/wuzhikun:/data  192.168.11.102:5000/cnvnator:1.0  cnvnator -unique -root /data/Project/Illumina_Trio/CNVnator/M625-0_CNV   -tree /data/Project/Illumina_Trio/mapping/M625-0/M625-0.bqsr.bam

```

output file:
```
$ ls -lh  ~/Project/Illumina_Trio/CNVnator/M625-0_CNV 
-rw-r--r-- 1 root root 1.0G May 31 16:15 /home/wuzhikun/Project/Illumina_Trio/CNVnator/M625-0_CNV

```

### step 2

```
$ docker run  -v /home/wuzhikun:/data  192.168.11.102:5000/cnvnator:1.0  cnvnator -unique -root /data/Project/Illumina_Trio/CNVnator/M625-0_CNV   -d  /data/database/genome/GRCh38/Chrs    -his 500 -chrom 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y 

Allocating memory ...
Done.
Calculating histograms with bin size of 500 for '1' ...
Making directory bin_500 ...
Making GC histogram for '1' ...
Done.

```

output file
```
$ ls -lh  ~/Project/Illumina_Trio/CNVnator/M625-0_CNV 
-rw-r--r-- 1 root root 1.1G May 31 16:33 /home/wuzhikun/Project/Illumina_Trio/CNVnator/M625-0_CNV

```


### step 3
```
$ docker run  -v /home/wuzhikun:/data  192.168.11.102:5000/cnvnator:1.0  cnvnator -unique -root /data/Project/Illumina_Trio/CNVnator/M625-0_CNV  -genome /data/database/genome/GRCh38/Homo_sapiens.GRCh38.dna.primary_assembly.fa  -stat 500


aking statistics for 1 ...
Making statistics for 2 ...
Making statistics for 3 ...
Making statistics for 4 ...
Making statistics for 5 ...
Making statistics for 6 ...
Making statistics for 7 ...
Making statistics for 8 ...
Making statistics for 9 ...
Making statistics for 10 ...
Making statistics for 11 ...
Making statistics for 12 ...
Making statistics for 13 ...
Making statistics for 14 ...
Making statistics for 15 ...
Making statistics for 16 ...
Making statistics for 17 ...
Making statistics for 18 ...
Making statistics for 19 ...
Making statistics for 20 ...
Making statistics for 21 ...
Making statistics for 22 ...
Making statistics for X ...
Making statistics for Y ...
Average RD per bin (1-22) is 91.3084 +- 15.5761 (before GC correction)
Average RD per bin (X,Y)  is 46.8913 +- 10.2851 (before GC correction)
Correcting counts by GC-content for '1' ...
Correcting counts by GC-content for '2' ...
Correcting counts by GC-content for '3' ...
Correcting counts by GC-content for '4' ...
Correcting counts by GC-content for '5' ...
Correcting counts by GC-content for '6' ...
Correcting counts by GC-content for '7' ...
Correcting counts by GC-content for '8' ...
Correcting counts by GC-content for '9' ...
Correcting counts by GC-content for '10' ...
Correcting counts by GC-content for '11' ...
Correcting counts by GC-content for '12' ...
Correcting counts by GC-content for '13' ...
Correcting counts by GC-content for '14' ...
Correcting counts by GC-content for '15' ...
Correcting counts by GC-content for '16' ...
Correcting counts by GC-content for '17' ...
Correcting counts by GC-content for '18' ...
Correcting counts by GC-content for '19' ...
Correcting counts by GC-content for '20' ...
Correcting counts by GC-content for '21' ...
Correcting counts by GC-content for '22' ...
Correcting counts by GC-content for 'X' ...
Correcting counts by GC-content for 'Y' ...
Making statistics for 1 after GC correction ...
Making statistics for 2 after GC correction ...
Making statistics for 3 after GC correction ...
Making statistics for 4 after GC correction ...
Making statistics for 5 after GC correction ...
Making statistics for 6 after GC correction ...
Making statistics for 7 after GC correction ...
Making statistics for 8 after GC correction ...
Making statistics for 9 after GC correction ...
Making statistics for 10 after GC correction ...
Making statistics for 11 after GC correction ...
Making statistics for 12 after GC correction ...
Making statistics for 13 after GC correction ...
Making statistics for 14 after GC correction ...
Making statistics for 15 after GC correction ...
Making statistics for 16 after GC correction ...
Making statistics for 17 after GC correction ...
Making statistics for 18 after GC correction ...
Making statistics for 19 after GC correction ...
Making statistics for 20 after GC correction ...
Making statistics for 21 after GC correction ...
Making statistics for 22 after GC correction ...
Making statistics for X after GC correction ...
Making statistics for Y after GC correction ...
Average RD per bin (1-22) is 89.0637 +- 13.7852 (after GC correction)
Average RD per bin (X,Y)  is 39.3075 +- 8.69979 (after GC correction)

```




### step 4
```
docker run  -v /home/wuzhikun:/data  192.168.11.102:5000/cnvnator:1.0  cnvnator -unique -root /data/Project/Illumina_Trio/CNVnator/M625-0_CNV  -genome /data/database/genome/GRCh38/Homo_sapiens.GRCh38.dna.primary_assembly.fa  -partition  500



Partitioning RD signal for 'Y' with bin size of 500 ...
Average RD per bin is 39.3075 +- 8.69979
Bin band is 2
Bin band is 3
Bin band is 4
Bin band is 5
Bin band is 6
Bin band is 7
Bin band is 8
Bin band is 10
Bin band is 12
Bin band is 14
Bin band is 16
Bin band is 20
Bin band is 24
Bin band is 28
Bin band is 32
Bin band is 40
Bin band is 48
Bin band is 56
Bin band is 64
Bin band is 80
Bin band is 96
Bin band is 112
Bin band is 128

```

### step 5
```
$ docker run  -v /home/wuzhikun:/data  192.168.11.102:5000/cnvnator:1.0  cnvnator -unique -root /data/Project/Illumina_Trio/CNVnator/M625-0_CNV  -genome /data/database/genome/GRCh38/Homo_sapiens.GRCh38.dna.primary_assembly.fa  -call  500 >  M625-0_CNV_raw
Assuming male individual!
Assuming male individual!

```


output file:
```
$ head  M625-0_CNV_raw
deletion	1:1-10000	10000	0	1.59373e-11	7.02612e-197	1.99216e-11	9.31063e-156	-1
duplication	1:93001-128500	35500	1.70899	0	4.20354e+08	0	4.68408e+08	0.549262
deletion	1:140001-155500	15500	0.609006	7.71466e-08	10873.5	1.26921e-05	54442.6	0.936985
deletion	1:168001-178500	10500	0.589362	0.00253492	61.9422	0.0819539	1787.25	0.960644
deletion	1:207501-258000	50500	0.0059667	3.15589e-12	0	3.28603e-12	0	0.989899
duplication	1:261501-281500	20000	1.78673	5.15506e-07	107568	3.58507e-06	298070	0.547258
deletion	1:295001-348000	53000	0.0408217	3.00703e-12	275.347	3.12495e-12	506.66	0.757914
deletion	1:358001-502000	144000	0.635718	1.10675e-12	2.871e+09	1.12234e-12	2.871e+09	0.967981
deletion	1:536001-586000	50000	0	3.18745e-12	0	3.32026e-12	0	1
deletion	1:670501-676500	6000	0.515772	0.0306213	4.06215e-06	6.72928	0.361839	0.760064

```


### step 6
```
$ perl /home/wuzhikun/anaconda3/pkgs/cnvnator-0.3.3-h79ee3db_1/bin/cnvnator2VCF.pl M625-0_CNV_raw  ~/database/genome/GRCh38/Chrs > M625-0_CNV.vcf


```


output file:
```
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  M625-0_CNV_raw
1       1       CNVnator_del_1  N       <DEL>   .       PASS    END=10000;SVTYPE=DEL;SVLEN=-10000;IMPRECISE;natorRD=0;natorP1=1.59373e-11;natorP2=7.02612e-197;natorP3=1.99216e-11;natorP4=9.31063e-156;natorQ0=-1      GT:CN   1/1:0
1       93001   CNVnator_dup_2  T       <DUP>   .       PASS    END=128500;SVTYPE=DUP;SVLEN=35500;IMPRECISE;natorRD=1.70899;natorP1=0;natorP2=4.20354e+08;natorP3=0;natorP4=4.68408e+08;natorQ0=0.549262
        GT:CN   ./1:2
1       140001  CNVnator_del_3  G       <DEL>   .       PASS    END=155500;SVTYPE=DEL;SVLEN=-15500;IMPRECISE;natorRD=0.609006;natorP1=7.71466e-08;natorP2=10873.5;natorP3=1.26921e-05;natorP4=54442.6;natorQ0=0.936985  GT:CN   0/1:1
1       168001  CNVnator_del_4  G       <DEL>   .       PASS    END=178500;SVTYPE=DEL;SVLEN=-10500;IMPRECISE;natorRD=0.589362;natorP1=0.00253492;natorP2=61.9422;natorP3=0.0819539;natorP4=1787.25;natorQ0=0.960644     GT:CN   0/1:1
1       207501  CNVnator_del_5  T       <DEL>   .       PASS    END=258000;SVTYPE=DEL;SVLEN=-50500;IMPRECISE;natorRD=0.0059667;natorP1=3.15589e-12;natorP2=0;natorP3=3.28603e-12;natorP4=0;natorQ0=0.989899     GT:CN   1/1:0

```