## Websites of Databases

### ARDB
```
## https://ardb.cbcb.umd.edu/
wget ftp://ftp.cbcb.umd.edu/pub/data/ARDB/ARDBflatFiles.tar.gz --no-check-certificate
wget ftp://ftp.cbcb.umd.edu/pub/data/ARDB/doc4ARDBflatFiles.pdf --no-check-certificate
wget ftp://ftp.cbcb.umd.edu/pub/data/ARDB/ardbAnno1.0.tar.gz --no-check-certificate
wget https://ardb.cbcb.umd.edu/ardbAnnoNotes.pdf --no-check-certificate
wget https://ardb.cbcb.umd.edu/tutorARDB.pdf --no-check-certificate
```


### carbon genes in uniprot
```
https://www.uniprot.org/uniprot/?query=carbon&sort=score
https://www.uniprot.org/uniprot/?query=carbon+cycle&sort=score
https://www.kegg.jp/kegg-bin/search_pathway_text?map=map&keyword=carbon&mode=1&viewImage=true
```

### [BioGRID](https://thebiogrid.org): Database of Protein, Chemical, and Genetic Interactions
```
wget https://downloads.thebiogrid.org/Download/BioGRID/Release-Archive/BIOGRID-3.4.160/BIOGRID-ALL-3.4.160.mitab.zip
wget https://downloads.thebiogrid.org/Download/BioGRID/Release-Archive/BIOGRID-3.4.160/BIOGRID-ALL-3.4.160.psi.zip
wget https://downloads.thebiogrid.org/Download/BioGRID/Release-Archive/BIOGRID-3.4.160/BIOGRID-ALL-3.4.160.psi25.zip
wget https://downloads.thebiogrid.org/Download/BioGRID/Release-Archive/BIOGRID-3.4.160/BIOGRID-ALL-3.4.160.tab.zip
wget https://downloads.thebiogrid.org/Download/BioGRID/Release-Archive/BIOGRID-3.4.160/BIOGRID-ALL-3.4.160.tab2.zip

```

### CDD
```
wget ftp://ftp.ncbi.nlm.nih.gov/pub/mmdb/cdd/cdd.info
wget ftp://ftp.ncbi.nlm.nih.gov/pub/mmdb/cdd/cdd.tar.gz
wget ftp://ftp.ncbi.nlm.nih.gov/pub/mmdb/cdd/cddannot.dat.gz
wget ftp://ftp.ncbi.nlm.nih.gov/pub/mmdb/cdd/cddid.tbl.gz
wget ftp://ftp.ncbi.nlm.nih.gov/pub/mmdb/cdd/cddid_all.tbl.gz
wget ftp://ftp.ncbi.nlm.nih.gov/pub/mmdb/cdd/cddmasters.fa.gz
```

### centifuge
```
wget ftp://ftp.ccb.jhu.edu/pub/infphilo/centrifuge/data/p_compressed+h+v.tar.gz
```

### chocophlan
```
wget http://huttenhower.sph.harvard.edu/humann2_data/chocophlan/chocophlan.tar.gz
```

### COG
```
wget ftp://ftp.ncbi.nih.gov/pub/COG/COG2014/data/Readme.201610.txt
wget ftp://ftp.ncbi.nih.gov/pub/COG/COG2014/data/cog2003-2014.csv
wget ftp://ftp.ncbi.nih.gov/pub/COG/COG2014/data/cognames2003-2014.tab
wget ftp://ftp.ncbi.nih.gov/pub/COG/COG2014/data/fun2003-2014.tab
wget ftp://ftp.ncbi.nih.gov/pub/COG/COG2014/data/genomes2003-2014.tab
wget ftp://ftp.ncbi.nih.gov/pub/COG/COG2014/data/prot2003-2014.fa.gz
wget ftp://ftp.ncbi.nih.gov/pub/COG/COG2014/data/prot2003-2014.gi2gbk.tab
wget ftp://ftp.ncbi.nih.gov/pub/COG/COG2014/data/prot2003-2014.tab
```

```
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Cdd_LE.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Cdd_NCBI_LE.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Cog_LE.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Kog_LE.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Pfam_LE.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Prk_LE.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Smart_LE.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Tigr_LE.tar.gz
wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/README
```

### dbCAN (CAZy)
```
## http://csbl.bmb.uga.edu/dbCAN/download.php
wget http://csbl.bmb.uga.edu/dbCAN/download/dbCAN-all-domains.txt
wget http://csbl.bmb.uga.edu/dbCAN/download/CAZyDB.07202017.fa
wget http://csbl.bmb.uga.edu/dbCAN/download/dbCAN-all-domains.txt.1e-5
```

### eggNOG
```
#http://eggnog.embl.de/version_3.0/downloads.html
#General Files
#wget http://eggnog.embl.de/version_3.0/data/downloads/species.v3.txt
#wget http://eggnog.embl.de/version_3.0/data/downloads/protein.aliases.v3.txt.gz
#wget http://eggnog.embl.de/version_3.0/data/downloads/fun.txt.gz
#wget http://eggnog.embl.de/version_3.0/data/downloads/UniProtAC2eggNOG.3.0.tsv.gz
#wget http://eggnog.embl.de/version_3.0/data/downloads/levels.txt
#wget http://eggnog.embl.de/version_3.0/data/downloads/sequences.v3.tar.gz
#wget http://eggnog.embl.de/version_3.0/data/downloads/sequences.raw.v3.tar.gz


#All COGs, KOGs, arCOGs and non-supervised orthologous groups and their proteins
wget http://eggnog.embl.de/version_3.0/data/downloads/all.members.tar.gz
#Bacteria non-supervised orthologous groups (bactNOG) and their proteins
wget http://eggnog.embl.de/version_3.0/data/downloads/bactNOG.members.txt.gz
#Clusters of orthologous groups (COG) and their proteins. For the complete set of orthologous groups in LUCA consider the NOGs as well.
wget http://eggnog.embl.de/version_3.0/data/downloads/COG.members.txt.gz
#Non-supervised orthologous groups (NOG) and their proteins. For the complete set of orthologous groups in LUCA consider the COGs as well.
wget http://eggnog.embl.de/version_3.0/data/downloads/NOG.members.txt.gz
#Eukaryote clusters of orthologous groups (KOG) and their proteins. For the complete set of eukaryotic orthologous groups consider the euNOGs as well.
wget http://eggnog.embl.de/version_3.0/data/downloads/KOG.members.txt.gz
#Vertebrates non-supervised orthologous groups (veNOG) and their proteins
wget http://eggnog.embl.de/version_3.0/data/downloads/veNOG.members.txt.gz




wget http://eggnog.embl.de/version_3.0/data/downloads/all.description.tar.gz
wget http://eggnog.embl.de/version_3.0/data/downloads/bactNOG.description.txt.gz
wget http://eggnog.embl.de/version_3.0/data/downloads/COG.description.txt.gz
wget http://eggnog.embl.de/version_3.0/data/downloads/NOG.description.txt.gz
wget http://eggnog.embl.de/version_3.0/data/downloads/KOG.description.txt.gz
wget http://eggnog.embl.de/version_3.0/data/downloads/veNOG.description.txt.gz
```

### eggnogdb
```
wget http://eggnogdb.embl.de/download/eggnog_4.5/eggnog-mapper-data/OG_fasta.tar.gz
wget http://eggnogdb.embl.de/download/eggnog_4.5/eggnog-mapper-data/eggnog4.clustered_proteins.fa.gz
wget http://eggnogdb.embl.de/download/eggnog_4.5/eggnog-mapper-data/eggnog.db.gz
wget http://eggnogdb.embl.de/download/eggnog_4.5/eggnog-mapper-data/og2level.tsv.gz
```

### GRCh38 germline variants
```
awk '{if($3 =='indel') {print $0}}' GRCh38.remap.all.germline.ucsc.gvf > GRCh38.remap.all.germline.ucsc.indel
```

### humann2

website
```
http://huttenhower.sph.harvard.edu/humann2_data/chocophlan/
http://huttenhower.sph.harvard.edu/humann2_data/uniprot/
http://huttenhower.sph.harvard.edu/humann2_data/uniprot/uniref_ec_filtered/
```

data
```
wget http://huttenhower.sph.harvard.edu/humann2_data/uniprot/uniref_annotated/uniref90_annotated.fasta.gz
wget http://huttenhower.sph.harvard.edu/humann2_data/uniprot/uniref_ec_filtered/uniref90_ec_filtered.tar.gz
```

diamond index
```
diamond makedb --in uniref90_annotated.fasta --db  uniref90_annotated.dmnd
```


### idmapping
```
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/idmapping.dat.example
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/idmapping.dat.gz
# wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/HUMAN_9606_idmapping.dat.gz

wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/ARATH_3702_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/CAEEL_6239_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/CHICK_9031_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/DANRE_7955_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/DICDI_44689_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/DROME_7227_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/ECOLI_83333_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/HUMAN_9606_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/MOUSE_10090_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/RAT_10116_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/RAT_10116_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/YEAST_559292_idmapping.dat.gz
wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/RELEASE.metalink
```


### IntAct
```
wget ftp://ftp.ebi.ac.uk/pub/databases/intact/current/all.zip
```

```
cut -f 1  intact.txt | cut -d ":" -f 1 | uniq | sort |uniq > intact_database
```


### interPro
```
### [InterPro in 2017—beyond protein family and domain annotations](http://nar.oxfordjournals.org/content/45/D1/D190)
### [interpro](https://www.ebi.ac.uk/interpro/download.html)
wget ftp://ftp.ebi.ac.uk/pub/software/unix/iprscan/5/5.28-67.0/interproscan-5.28-67.0-64-bit.tar.gz
wget ftp://ftp.ebi.ac.uk/pub/databases/interpro/entry.list
wget ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz
wget ftp://ftp.ebi.ac.uk/pub/databases/interpro/match_complete.xml.gz
wget ftp://ftp.ebi.ac.uk/pub/databases/interpro/uniparc_match.tar.gz
wget ftp://ftp.ebi.ac.uk/pub/databases/interpro/protein2ipr.dat.gz
wget ftp://ftp.ebi.ac.uk/pub/databases/interpro/ParentChildTreeFile.txt
wget ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro2go
wget ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt
```

### JASPAR
```
wget http://jaspar.genereg.net/download/CORE/JASPAR2018_CORE_plants_non-redundant_pfms_jaspar.zip
wget http://jaspar.genereg.net/download/CORE/JASPAR2018_CORE_plants_non-redundant_pfms_meme.zip
wget http://jaspar.genereg.net/download/CORE/JASPAR2018_CORE_plants_non-redundant_pfms_transfac.zip

```


### KOBAS
```
python ~/github/zkwu/kcBSA/pipeline/downKOBASdb.py kobas.html /home/wzk/test/kobas
nohup python ~/github/zkwu/kcBSA/pipeline/downKOBASdb.py kobas.html /home/wzk/test/kobas >temp 2>&1 &
```

```
#2017/2/10 上午8:00:00
#http://kobas.cbi.pku.edu.cn/download.php
wget ftp://ftp.cbi.pku.edu.cn/pub/KOBAS_3.0_DOWNLOAD/species_abbr.txt
wget ftp://ftp.cbi.pku.edu.cn/pub/KOBAS_3.0_DOWNLOAD/README.txt
wget ftp://ftp.cbi.pku.edu.cn/pub/KOBAS_3.0_DOWNLOAD/seq_pep/ko.pep.fasta.gz
wget ftp://ftp.cbi.pku.edu.cn/pub/KOBAS_3.0_DOWNLOAD/sqlite3/ko.db.gz
wget http://kobas.cbi.pku.edu.cn/download_file.php?type=sqlite3&filename=organism.db.gz

```


### metaphlan2
```
wget https://bitbucket.org/biobakery/metaphlan2/downloads/mpa_v20_m200_marker_info.txt.bz2
wget https://bitbucket.org/biobakery/metaphlan2/downloads/mpa_v20_m200.tar
```


### NCBI
```
wget ftp://ftp.ncbi.nih.gov/blast/db/FASTA/nr.gz
wget ftp://ftp.ncbi.nih.gov/blast/db/FASTA/nr.gz.md5
wget ftp://ftp.ncbi.nih.gov/blast/db/FASTA/nt.gz
wget ftp://ftp.ncbi.nih.gov/blast/db/FASTA/nt.gz.md5
```

```
nohup diamond makedb --in nr.fasta -d nr  &
diamond blastp --db /home/wzk/database/NCBI/nr.dmnd  --query /home/wzk/BSA_wheat/Candidate/all_gene.fasta --threads 20 --evalue 0.00001 --max-target-seqs 1 --out /home/wzk/BSA_wheat/Candidate/all_gene_blast.txt
```


#### download local database
```
#!/usr/bin/env bash

for ((i=10; i<=59; i++))
do 
#    wget ftp://ftp.ncbi.nih.gov/blast/db/nt.$i.tar.gz 
#  tar -zxf nt.$1.tar.gz
    wget ftp://ftp.ncbi.nih.gov/blast/db/nt.$i.tar.gz.md5
    done

    ```


    ### NCyc
    ```
    https://codeload.github.com/qichao1984/NCyc/zip/master
    ```

    ### P450 in uniprot
    ```
    http://p450.riceblast.snu.ac.kr/download.php?a=download
    ```


    ### Pfam

    ```
    #wget ftp://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam31.0/Pfam-A.full.uniprot.gz
    wget ftp://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam31.0/uniprot.gz
    ```

    ### RNAmmer
    ```
    wget http://www.cbs.dtu.dk/public/RNAmmer-1.2/data/rnammer-1.2.fsa.gz
    wget http://www.cbs.dtu.dk/public/RNAmmer-1.2/data/rnammer-1.2.gff.gz
    wget http://www.cbs.dtu.dk/public/RNAmmer-1.2/data/rnammer-1.2.hmm.gz
    wget http://www.cbs.dtu.dk/public/RNAmmer-1.2/data/rnammer-1.2.md5.gz
    ```

    ### rrnDB

    ```
    wget https://rrndb.umms.med.umich.edu/static/download/rrnDB-5.4_pantaxa_stats_RDP.tsv.zip
    wget https://rrndb.umms.med.umich.edu/static/download/rrnDB-5.4_pantaxa_stats_NCBI.tsv.zip
    wget https://rrndb.umms.med.umich.edu/static/download/rrnDB-5.4_16S_rRNA.fasta.zip
    wget https://rrndb.umms.med.umich.edu/static/download/rrnDB-5.4.tsv.zip

    ```

    ### SILVA
    ```
    wget https://www.arb-silva.de/fileadmin/silva_databases/release_132/Exports/SILVA_132_SSURef_tax_silva.fasta.gz
    wget https://www.arb-silva.de/fileadmin/silva_databases/release_132/Exports/SILVA_132_SSURef_tax_silva.fasta.gz.md5
    wget https://www.arb-silva.de/fileadmin/silva_databases/release_132/Exports/SILVA_132_SSURef_Nr99_tax_silva.fasta.gz
    wget https://www.arb-silva.de/fileadmin/silva_databases/release_132/Exports/SILVA_132_SSURef_Nr99_tax_silva.fasta.gz.md5
    ```


    ### SPRINT

    ```
    wget http://sprint.software/static/source/uploads/dbrep/dm6_repeat.txt
    wget http://sprint.software/static/source/uploads/dbrep/hg38_repeat.txt
    wget http://sprint.software/static/source/uploads/dbrep/mm10_repeat.txt
    wget http://sprint.software/static/source/uploads/dbrep/hg19_repeat.txt
    wget http://sprint.software/static/source/uploads/dbrep/mm9_repeat.txt
    wget http://sprint.software/static/source/uploads/dbrep/panTro3_repeat.txt
    wget http://sprint.software/static/source/uploads/dbrep/MacaM_Rhesus_v7_repeat.txt
    wget http://sprint.software/static/source/uploads/dbrep/ce10_repeat.txt
    ```

    ### taxonomy
    ```
    wget ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz
    wget ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz.md5
    wget ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump_readme.txt
    ```

    ### uniprot
    ```
    wget wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz
    wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_trembl.fasta.gz

    ```


    ### VFDB
    ```
    ## http://www.mgc.ac.cn/VFs/download.htm
    wget http://www.mgc.ac.cn/VFs/Down/VFs.xls.gz
    wget http://www.mgc.ac.cn/VFs/Down/Comparative_tables_from_VFDB.tar.gz
    wget http://www.mgc.ac.cn/VFs/Down/VFDB_setA_nt.fas.gz
    wget http://www.mgc.ac.cn/VFs/Down/VFDB_setA_pro.fas.gz
    wget http://www.mgc.ac.cn/VFs/Down/VFDB_setB_nt.fas.gz
    wget http://www.mgc.ac.cn/VFs/Down/VFDB_setB_pro.fas.gz
    ```

