# Nanopore quality control

## [nanofilt](https://github.com/wdecoster/nanofilt)
## [NanoQC](https://github.com/wdecoster/nanoQC)
## [Filtlong](https://github.com/rrwick/Filtlong): quality filtering tool for long reads
## [minion_qc](https://github.com/roblanf/minion_qc): Quality control for MinION sequencing data
## [NanoLyse](https://github.com/wdecoster/nanolyse)


## nanostat

### install nanostat
```
$ conda install -c bioconda nanostat
```

### parameters
```
$ NanoStat --help
usage: NanoStat [-h] [-v] [-o OUTDIR] [-p PREFIX] [-n NAME] [-t N]
                [--barcoded] [--readtype {1D,2D,1D2}]
                (--fastq file [file ...] | --fasta file [file ...] | --summary file [file ...] | --bam file [file ...])

Calculate statistics of long read sequencing dataset.

General options:
  -h, --help            show the help and exit
  -v, --version         Print version and exit.
  -o, --outdir OUTDIR   Specify directory for output, only in combination with -n.
  -p, --prefix PREFIX   Specify an optional prefix to be used for the output file.
  -n, --name NAME       Specify a filename/path for the output, stdout is the default.
  -t, --threads N       Set the allowed number of threads to be used by the script.

Input options.:
  --barcoded            Use if you want to split the summary file by barcode
  --readtype {1D,2D,1D2}
                        Which read type to extract information about from summary. Options are 1D, 2D,
                        1D2

Input data sources, one of these is required.:
  --fastq file [file ...]
                        Data is in one or more (compressed) fastq file(s).
  --fasta file [file ...]
                        Data is in one or more (compressed) fasta file(s).
  --summary file [file ...]
                        Data is in one or more (compressed) summary file(s)generated by albacore.
  --bam file [file ...]
                        Data is in one or more sorted bam file(s).

EXAMPLES:
  NanoStat --fastq reads.fastq.gz --outdir statreports
  NanoStat --summary sequencing_summary1.txt sequencing_summary2.txtsequencing_summary3.txt --readtype 1D2
  NanoStat --bam alignment.bam alignment2.bam
```

### run NanoStat
```
$ NanoStat --threads 5  --outdir /disk1/Project/KC2018-C128/PRJEB26791/raw/ERR2631603 --name call --fastq /disk1/Project/KC2018-C128/PRJEB26791/raw/ERR2631603.fastq.gz
```

output file:
```
$ cat  /disk1/Project/KC2018-C128/PRJEB26791/raw/ERR2631603/call
General summary:        
Mean read length:              16,018.1
Mean read quality:                  7.8
Median read length:            12,627.0
Median read quality:                8.7
Number of reads:               10,000.0
Read length N50:               26,852.0
Total bases:              160,180,579.0
Number, percentage and megabases of reads above quality cutoffs
>Q5:    8163 (81.6%) 153.8Mb
>Q7:    7438 (74.4%) 142.3Mb
>Q10:   918 (9.2%) 15.8Mb
>Q12:   0 (0.0%) 0.0Mb
>Q15:   0 (0.0%) 0.0Mb
Top 5 highest mean basecall quality scores and their read lengths
1:  11.5 (5423)
2:  11.2 (825)
3:  11.2 (17797)
4:  11.0 (7417)
5:  10.9 (5073)
Top 5 longest reads and their mean basecall quality score
1:  104171 (7.8)
2:  91897 (7.6)
3:  90216 (7.7)
4:  89826 (7.4)
5:  89199 (8.7)

```





## [NanoQC](https://github.com/wdecoster/nanoQC)

### install NanoQC
```
$ conda install -c bioconda nanoqc
```

### NanoQC parameters

```
$ nanoQC --help
usage: nanoQC [-h] [-v] [-o OUTDIR] [-l MINLEN] fastq

Investigate nucleotide composition and base quality.

positional arguments:
  fastq                 Reads data in fastq.gz format.

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         Print version and exit.
  -o OUTDIR, --outdir OUTDIR
                        Specify directory in which output has to be created.
  -l MINLEN, --minlen MINLEN
                        Filters the reads on a minimal length of the given
                        range. Also plots the given length/2 of the begin and
                        end of the reads.
```


### run NanoQC

```
$ nanoQC --outdir nanoQC  ERR2631603.fastq.gz
```

output files:
```
$ tree nanoQC
nanoQC
├── nanoQC.html
└── NanoQC.log

```

## [nanofilt](https://github.com/wdecoster/nanofilt)
### install nanofilt
```
$ conda install -c bioconda nanofilt
```


### parameters
```
$ NanoFilt --help
usage: NanoFilt [-h] [-v] [--logfile LOGFILE] [-l LENGTH]
                [--maxlength MAXLENGTH] [-q QUALITY] [--minGC MINGC]
                [--maxGC MAXGC] [--headcrop HEADCROP] [--tailcrop TAILCROP]
                [-s SUMMARY] [--readtype {1D,2D,1D2}]

Perform quality and/or length and/or GC filtering of (long read) fastq data.           Reads on stdin.

General options:
  -h, --help            show the help and exit
  -v, --version         Print version and exit.
  --logfile LOGFILE     Specify the path and filename for the log file.

Options for filtering reads on.:
  -l LENGTH, --length LENGTH
                        Filter on a minimum read length
  --maxlength MAXLENGTH
                        Filter on a maximum read length
  -q QUALITY, --quality QUALITY
                        Filter on a minimum average read quality score
  --minGC MINGC         Sequences must have GC content >= to this. Float between 0.0 and 1.0. Ignored if
                        using summary file.
  --maxGC MAXGC         Sequences must have GC content <= to this. Float between 0.0 and 1.0. Ignored if
                        using summary file.

Options for trimming reads.:
  --headcrop HEADCROP   Trim n nucleotides from start of read
  --tailcrop TAILCROP   Trim n nucleotides from end of read

Input options.:
  -s SUMMARY, --summary SUMMARY
                        Use summary file for quality scores
  --readtype {1D,2D,1D2}
                        Which read type to extract information about from summary. Options are 1D, 2D or
                        1D2

EXAMPLES:
  gunzip -c reads.fastq.gz | NanoFilt -q 10 -l 500 --headcrop 50 | minimap2 genome.fa - | samtools sort -O BAM -@24 -o alignment.bam -
  gunzip -c reads.fastq.gz | NanoFilt -q 12 --headcrop 75 | gzip > trimmed-reads.fastq.gz
  gunzip -c reads.fastq.gz | NanoFilt -q 10 | gzip > highQuality-reads.fastq.gz

```

```
$ gunzip -c ERR2631603.fastq.gz | NanoFilt --quality  5 --length 500 --headcrop 40 --tailcrop 20 | gzip > highQuality.fastq.gz
```


```
$ grep -c '^@' ERR2631603.fastq
10000

$ grep -c '^@' highQuality.fastq
8054

```


## nanoplot

###  install nanoplot
```
$ conda install -c bioconda nanoplot
```

### parameters
```
$ NanoPlot --help
usage: NanoPlot [-h] [-v] [-t THREADS] [--verbose] [--store] [--raw]
                [-o OUTDIR] [-p PREFIX] [--maxlength N] [--minlength N]
                [--drop_outliers] [--downsample N] [--loglength]
                [--percentqual] [--alength] [--minqual N] [--runtime_until N]
                [--readtype {1D,2D,1D2}] [--barcoded] [-c COLOR]
                [-f {eps,jpeg,jpg,pdf,pgf,png,ps,raw,rgba,svg,svgz,tif,tiff}]
                [--plots [{kde,hex,dot,pauvre} [{kde,hex,dot,pauvre} ...]]]
                [--listcolors] [--no-N50] [--N50] [--title TITLE]
                [--font_scale FONT_SCALE] [--dpi DPI]
                (--fastq file [file ...] | --fasta file [file ...] | --fastq_rich file [file ...] | --fastq_minimal file [file ...] | --summary file [file ...] | --bam file [file ...] | --cram file [file ...] | --pickle pickle)

CREATES VARIOUS PLOTS FOR LONG READ SEQUENCING DATA.

General options:
  -h, --help            show the help and exit
  -v, --version         Print version and exit.
  -t, --threads THREADS
                        Set the allowed number of threads to be used by the script
  --verbose             Write log messages also to terminal.
  --store               Store the extracted data in a pickle file for future plotting.
  --raw                 Store the extracted data in tab separated file.
  -o, --outdir OUTDIR   Specify directory in which output has to be created.
  -p, --prefix PREFIX   Specify an optional prefix to be used for the output files.

Options for filtering or transforming input prior to plotting:
  --maxlength N         Drop reads longer than length specified.
  --minlength N         Drop reads shorter than length specified.
  --drop_outliers       Drop outlier reads with extreme long length.
  --downsample N        Reduce dataset to N reads by random sampling.
  --loglength           Logarithmic scaling of lengths in plots.
  --percentqual         Use qualities as theoretical percent identities.
  --alength             Use aligned read lengths rather than sequenced length (bam mode)
  --minqual N           Drop reads with an average quality lower than specified.
  --runtime_until N     Only take the N first hours of a run
  --readtype {1D,2D,1D2}
                        Which read type to extract information about from summary. Options are 1D, 2D,
                        1D2
  --barcoded            Use if you want to split the summary file by barcode

Options for customizing the plots created:
  -c, --color COLOR     Specify a color for the plots, must be a valid matplotlib color
  -f, --format {eps,jpeg,jpg,pdf,pgf,png,ps,raw,rgba,svg,svgz,tif,tiff}
                        Specify the output format of the plots.
  --plots [{kde,hex,dot,pauvre} [{kde,hex,dot,pauvre} ...]]
                        Specify which bivariate plots have to be made.
  --listcolors          List the colors which are available for plotting and exit.
  --no-N50              Hide the N50 mark in the read length histogram
  --N50                 Show the N50 mark in the read length histogram
  --title TITLE         Add a title to all plots, requires quoting if using spaces
  --font_scale FONT_SCALE
                        Scale the font of the plots by a factor
  --dpi DPI             Set the dpi for saving images

Input data sources, one of these is required.:
  --fastq file [file ...]
                        Data is in one or more default fastq file(s).
  --fasta file [file ...]
                        Data is in one or more fasta file(s).
  --fastq_rich file [file ...]
                        Data is in one or more fastq file(s) generated by albacore, MinKNOW or guppy
                        with additional information concerning channel and time.
  --fastq_minimal file [file ...]
                        Data is in one or more fastq file(s) generated by albacore, MinKNOW or guppy
                        with additional information concerning channel and time. Is extracted swiftly
                        without elaborate checks.
  --summary file [file ...]
                        Data is in one or more summary file(s) generated by albacore or guppy.
  --bam file [file ...]
                        Data is in one or more sorted bam file(s).
  --cram file [file ...]
                        Data is in one or more sorted cram file(s).
  --pickle pickle       Data is a pickle file stored earlier.

EXAMPLES:
    Nanoplot --summary sequencing_summary.txt --loglength -o summary-plots-log-transformed
    NanoPlot -t 2 --fastq reads1.fastq.gz reads2.fastq.gz --maxlength 40000 --plots hex dot
    NanoPlot --color yellow --bam alignment1.bam alignment2.bam alignment3.bam --downsample 10000

```


### run NanoPlot
```
$ NanoPlot --threads 4 --outdir NanoPlot --prefix ERR2631603 --fastq ERR2631603.fastq.gz --loglength --format pdf --plots hex dot
```

output files:
```
NanoPlot
├── ERR2631603HistogramReadlength.pdf
├── ERR2631603LengthvsQualityScatterPlot_dot.pdf
├── ERR2631603LengthvsQualityScatterPlot_hex.pdf
├── ERR2631603LogTransformed_HistogramReadlength.pdf
├── ERR2631603NanoPlot_20190115_0532.log
├── ERR2631603NanoPlot-report.html
├── ERR2631603NanoStats.txt
├── ERR2631603Weighted_HistogramReadlength.pdf
├── ERR2631603Weighted_LogTransformed_HistogramReadlength.pdf
└── ERR2631603Yield_By_Length.pdf
```


## [NanoLyse](https://github.com/wdecoster/nanolyse)

Remove reads mapping to the lambda phage genome from a fastq file.

### install NanoLyse
```
$ conda install -c bioconda nanolyse
```

```

$ NanoLyse --help
usage: NanoLyse [-h] [-v] [-r REFERENCE] [--logfile LOGFILE]

Remove reads mapping to the lambda genome. Reads fastq from stdin and writes to stdout.

optional arguments:
  -r, --reference REFERENCE
                        Specify a reference fasta file against which to filter.
  --logfile LOGFILE     Specify the path and filename for the log file.

General options:
  -h, --help            show the help and exit
  -v, --version         Print version and exit.

EXAMPLES:
    gunzip -c reads.fastq.gz | NanoLyse | gzip > reads_without_lambda.fastq.gz
    gunzip -c reads.fastq.gz | NanoLyse | NanoFilt -q 12 | gzip > filtered_reads_without_lambda.fastq.gz
    gunzip -c reads.fastq.gz | NanoLyse --reference mygenome.fa.gz | gzip > reads_without_mygenome.fastq.gz
```

### download reference data of phage lambda

Enterobacteria phage lambda, complete genome

GenBank: J02459.1
```
https://www.ncbi.nlm.nih.gov/nuccore/J02459.1?report=fasta
```

### run NanoLyse

```
$ gunzip -c ERR2631603.fastq.gz | NanoLyse --reference /home/wzk/github/NanoSV/data/Enterobacteria_phage_lambda.fasta  > reads_without_lambda.fastq

NanoLyse: removed 0 reads.
```

